src/mds/Server.cc

```diff
@@ -6252,6 +6260,119 @@ void Server::handle_client_removexattr(MDRequestRef& mdr)
   journal_and_reply(mdr, cur, 0, le, new C_MDS_inode_update_finish(this, mdr, cur));
 }
 
+void Server::handle_client_getvxattr(MDRequestRef& mdr)
+{
+  const cref_t<MClientRequest> &req = mdr->client_request;
+  string name(req->get_path2());
+
+  // is a ceph virtual xattr?
+  if (!is_ceph_vxattr(name)) {
+    respond_to_request(mdr, -CEPHFS_EINVAL);
+    return;
+  }
+
+  CInode *cur = try_get_auth_inode(mdr, req->get_filepath().get_ino());
+  if (!cur) {
+    return;
+  }
+  // so far we need all these locks for getting
+  // 1. dir layout and
+  // 2. dir pinning policies
+  MutationImpl::LockOpVec lov;
+  lov.add_rdlock(&cur->xattrlock);
+  lov.add_rdlock(&cur->snaplock);
+  lov.add_rdlock(&cur->policylock);
+  if (!mds->locker->acquire_locks(mdr, lov))
+    return;
+
+  if (!cur->is_dir()) {
+    respond_to_request(mdr, -CEPHFS_ENOTDIR);
+    return;
+  }
+
+  int r = 0;
+  ceph::bufferlist bl;
+  // handle these vxattrs
+  if (name.substr(0, 15) == "ceph.dir.layout"sv) {
+    std::string layout_field;
+
+    if (name.length() > 15)
+      layout_field = name.substr(15);
+
+    struct layout_xattr_info_t {
+      const file_layout_t& layout;
+      const std::string    inheritance_status;
+
+      layout_xattr_info_t(const file_layout_t& l, const std::string& is)
+        : layout(l), inheritance_status(is) { }
+    };
+
+    auto get_inherited_layout = [&](CInode *cur) -> layout_xattr_info_t {
+      auto orig_in = cur;
+      while (cur) {
+        if (cur->get_projected_inode()->has_layout()) {
+          if (cur == orig_in) {
+            // we've found a new layout at this inode
+            return {cur->get_projected_inode()->layout, "new"s};
+          } else {
+            return {cur->get_projected_inode()->layout, "inherited"s};
+          }
+        }
+
+        if (cur->is_root())
+          break;
+
+        cur = cur->get_parent_dir()->get_inode();
+      }
+      return {mdcache->default_file_layout, "default"s};
+    };
+
+    const auto& [layout, inheritance] = get_inherited_layout(cur);
+
+    if (layout_field == ""sv) {
+      layout.encode(bl, (uint64_t)-1);
+      encode(inheritance, bl);
+    } else if (layout_field == ".stripe_unit"sv) {
+      encode(layout.stripe_unit, bl);
+    } else if (layout_field == ".stripe_count"sv) {
+      encode(layout.stripe_count, bl);
+    } else if (layout_field == ".object_size"sv) {
+      encode(layout.object_size, bl);
+    } else if (layout_field == ".pool"sv) {
+      encode(layout.pool_id, bl);
+    } else if (layout_field == ".pool_namespace"sv) {
+      encode(layout.pool_ns, bl);
+    } else {
+      r = -CEPHFS_EINVAL;
+    }
+  } else if (name.substr(0, 12) == "ceph.dir.pin"sv) {
```

Comment:
This vxattr should return a text string. The whole point of `getvxattr` is so that we don't need to update the clients everytime we want to add one from the MDS side. If the client has to do binary processing on the PRC, that defeats the purpose, does it not?

Reply:
Sorry that mentioned the word 'encoding' at all. I just meant 'binary' and not 'binary encoding'
Wouldn't plain text transmission suffice, rather than JSON ?
What would be the input format to `setxattr` ? JSON as well ? like wrapping things in `{` and `}`
Just thinking about the input and output format consistency.

Reply:
For better or worse, JSON is becoming the common format for all ceph related introspection commands. It is not that hard to manipulate with `jq`.

When it makes sense, a raw value can be encoded by the `getvxattr` RPC (like `ceph.dir.pin.random`) but for compound values like `ceph.dir.layout`, JSON makes the most sense.